src/loaders/api/register-api.js

```diff
@@ -0,0 +1,108 @@
+    handle('api-pve', [
+      /** Handles the rum response when finished */
+      (data) => {
+        try {
+          clearTimeout(timeout)
+          const entityGuid = data.app.agents?.[0].entityGuid
+          if (!entityGuid) throw new Error('No target entity guid returned from PVE call')
+          target.entityGuid = entityGuid
+          resolve(data)
+        } catch (err) {
+          reject(err)
+        }
+      },
+      /** custom attributes to send with the rum call */
+      attrs,
+      /** target to send the rum call to (we dont have the entityguid yet in PVE call, so must supply a direct obj) */
+      target
+    ], undefined, FEATURE_NAMES.pageViewEvent, agentRef.ee)
+  })
+
+  /**
+     * The reporter method that will be used to report the data to the container agent's API method.
+     * If the external.capture_registered_data configuration value is set to true, the data will be reported to BOTH the container and the external target
+     * @param {*} methodToCall the container agent's API method to call
+     * @param {*} args the arguments to supply to the container agent's API method
+     * @param {string} targetEntityGuid the target entity guid, which looks up the target to report the data to from the entity manager. If undefined, will report to the container agent's target.
+     * @returns
+     */
+  const report = (methodToCall, args, target) => {
+    /** set the timestamp before the async part of waiting for the rum response for better accuracy */
+    const timestamp = now()
+    handle(SUPPORTABILITY_METRIC_CHANNEL, [`API/register/${methodToCall.name}/called`], undefined, FEATURE_NAMES.metrics, agentRef.ee)
+    waitForRumResponse.then((rumResponse) => {
+      // target should be decorated with entityGuid by the rum resp at this point
+      if (methodToCall === handlers.log && !(target.entityGuid && rumResponse.log)) return // not enough metadata <or> was not sampled
+      if (agentRef.init.external.capture_registered_data) { methodToCall(...args, undefined, timestamp) } // also report to container by providing undefined target
+      methodToCall(...args, target.entityGuid, timestamp) // report to target
+    }).catch(err => {
+      warn(49, err)
+    })
+  }
+  return {
+    addPageAction: (name, attributes = {}) => report(handlers.addPageAction, [name, { ...attrs, ...attributes }], target),
+    log: (message, options = {}) => report(handlers.log, [message, { ...options, customAttributes: { ...attrs, ...(options.customAttributes || {}) } }], target),
+    noticeError: (error, attributes = {}) => report(handlers.noticeError, [error, { ...attrs, ...attributes }], target),
+    setApplicationVersion: (value) => {
+      attrs['application.version'] = value
+    },
+    setCustomAttribute: (key, value) => {
+      attrs[key] = value
+    },
+    setUserId: (value) => {
+      attrs['enduser.id'] = value
+    },
+    /** metadata */
+    metadata: {
+      customAttributes: attrs,
+      target
+    }
```

Comment:
Maybe this should return a registeredEntity object? Some of these are getting defined multiple times, through placeholder and through actual implementation for both the main agent and MFE, and it's getting hard to keep track of which is which and from where.

Reply:
08b37627